routing/router.go

```diff
@@ -1916,13 +1868,11 @@ func (r *ChannelRouter) processSendError(paySession *paymentSession,
 	// we'll prune the channel in both directions and
 	// continue with the rest of the routes.
 	case *lnwire.FailPermanentChannelFailure:
-		paySession.ReportEdgeFailure(&EdgeLocator{
-			ChannelID: failedEdge.ChannelID,
-			Direction: 0,
-		})
-		paySession.ReportEdgeFailure(&EdgeLocator{
-			ChannelID: failedEdge.ChannelID,
-			Direction: 1,
+		paySession.ReportEdgeFailure(failedEdge)
```

Comment:
interesting that a permanent failure would have a half-life wrt to success probability. down the road are we planning to make probabilities more failure-specific?

for example, permanent failures could update the `nodeHistory` to then always report the probability as 0 for that channel

another thing that comes to mind is that a bandwidth failure for 100000 sats may cause a channel to be ignored when trying to route 1 sat payment directly after. seems we would need to track the size of the payment that generated the failure (at least in this case).

Reply:
thinking more on the bandwidth failures, i think we would only need to track the minimum payment size that caused the failure. could then define a piecewise probability function:
```
Pr[success] = 1               if amt < min_failed_amt
Pr[success] = 1 - 2^exp       if amt >= min_failed_amt
```

Reply:
Yes, so the question is how permanent a permanent failure is. Do we want to take into account that this result could have been sent because of a bug and give another chance?

For this PR, the probability calculation is very basic. I plan to do a follow up with much better interpretation of failures. Not only the permanent failures, but also other errors for which the error is currently not attributed to the right channel. 

The main goal for this pr is to make mission control global and have it remember things longer than it used to do, without regression.

Your comment about bandwidth failures is valid in that regard. Previously we tracked those failures only locally for a single payment. All routes for a payment used roughly the same amount (ignoring fees). Now we may prune channels because we tried a high amount, while they may still work for a low amount (in the next payment).

I think there are two ways of addressing this:

1. In line with your suggestion. Try to factor the amount into the probability. I like your idea because it is simple and will work almost the same as in master. For concurrent payments there will be interference between the different amounts probably, but I feel that is ok for now. Later we could go with multiple segments

2. Be more aggressive and expect nodes to keep their channels balanced. Only pick channels during path finding for which the payment amount is <30% of the channel capacity and immediately prune the channel if it fails.

Reply:
Also what could happen is that if we decide to be more relaxed about balance issues, nodes may just replace their actual error by a temp channel failure. To get penalized less. I tend to be cautious with coupling too specific behaviour to a particular fail code.

Reply:
Discussed with @Roasbeef as well. Option 2 is likely too early for the network in its current state.

Reply:
> Do we want to take into account that this result could have been sent because of a bug and give another chance?

Typically this would indicate the channel has already closed (via timeout resolutions). I suppose it doesn't matter too much, by the time the penalty has decayed enough to make the route attractive, we will have likely pruned the channel from our graph.

That being said, do we also want to prune mission control state when pruning the graph?

> Also what could happen is that if we decide to be more relaxed about balance issues, nodes may just replace their actual error by a temp channel failure. To get penalized less. I tend to be cautious with coupling too specific behaviour to a particular fail code.

Good point, we can never be 100% certain. If we want to treat all of the errors similarly, then I suppose we should track the failed payment amount for all failure types and always apply the piece-wise heuristic?

> Option 2 is likely too early for the network in its current state.

Agreed. I'd be in favor of seeing what option 1 looks like so that we maintain a close-ish behavior to current master

Reply:
Implemented option 1

Reply:
>That being said, do we also want to prune mission control state when pruning the graph?

I think I am more comfortable with pruning mission control independently. Just take a max age and remove when it is exceeded.

Also, observations may still be valuable after the channel is pruned. It indicates how the node behaved historically.

In this PR, mission control state grows in memory. I don't think it is a problem, as it it bounded by the total number of channels and nodes in the graph.

In the follow up where persistence is added, I will add pruning.